### PR TITLE
[FIX] hr,hr_contract,resource: allocate hours before resource creation

### PR DESCRIPTION
**Pull request by auon-odoo**
Steps :
\- On a new DB (no demo data). Say created at 15:00. \- Go to planning.
\- Add a shift for today.
\- Set the dates from 08:00 to 17:00.

Issue :
The 'Allocated Time' = 02:00.

Cause :
We used to count from the resource creation min to the departure max. So, in this case, not from 08:00, but from 15:00.
While this might seem logical, it confuses users in onboarding.

Fix :
Calculate the whole time, regardless of the resource lifespan.

Notes :
\- Similar issues solved with this commit :
	> Once the shift validated, the avatar progress bar uses
the same allocated time.
	> Same in Project Task gantt view.
\- When contract is installed, if the resource does not have a contract, the Allocated Time = 0. Now, it is calculated from the resource calendar.

task-2983993

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/hr_contract/models/resource_resource.py [hunk 1]
```diff
@@ -15,7 +15,23 @@ def _get_calendars_validity_within_period(self, start, end, default_company=None
...
         calendars_within_period_per_resource = defaultdict(lambda: defaultdict(Intervals))  # keys are [resource id:integer][calendar:self.env['resource.calendar']]
-        resource_without_contract = self.filtered(lambda r: not r.employee_id or r.employee_id.employee_type not in ['employee', 'student'])
+        employee_ids_with_active_contracts = {
+            contract['employee_id'][0] for contract in
+            self.env['hr.contract']._read_group(
+                domain=[
+                    ('employee_id', 'in', self.employee_id.ids),
+                    '|', ('state', '=', 'open'),
+                    '|', ('state', '=', 'close'),
+                         '&', ('state', '=', 'draft'), ('kanban_state', '=', 'done')
```
Comment (MissingNoShiny) [inline addons/hr_contract/models/resource_resource.py:25]: Is the contract really active if it's expired or in draft?

File: addons/resource/models/resource.py [hunk 2]
```diff
@@ -142,9 +142,9 @@ def _merge(self, other, difference):
 def sum_intervals(intervals):
     """ Sum the intervals duration (unit : hour)"""
     return sum(
-        (stop - start).total_seconds() / 3600
+        (stop - start).total_seconds()
         for start, stop, meta in intervals
-    )
+    ) / 3600
```
Comment (MissingNoShiny) [inline addons/resource/models/resource.py:147]: What is the purpose of this change?

**Review by MissingNoShiny — COMMENTED** [review]
Hello, thanks for you work, I left a few comments :slightly_smiling_face:

File: addons/resource/models/resource.py (hunk 2 above)
Comment (auon-odoo) [reply addons/resource/models/resource.py:147]: performance?

File: addons/hr_contract/models/resource_resource.py (hunk 1 above)
Comment (auon-odoo) [reply addons/hr_contract/models/resource_resource.py:25]: 'employee_ids_with_active_contracts' stands for 'employee_ids_that_have_or_had_an_active_contract'.
Thought it was a bit lighter :sweat_smile:

Comment (MissingNoShiny) [reply addons/hr_contract/models/resource_resource.py:25]: I see, thanks, maybe you could add a comment then, because it's confusing as is

File: addons/resource/models/resource.py (hunk 2 above)
Comment (MissingNoShiny) [reply addons/resource/models/resource.py:147]: I know 16.0 is not really stable yet, but we should still keep changes to a minimum. Besides, this change wouldn't have a meaningful effect on performance
